src/Bounds.cpp

```diff
@@ -1235,8 +1235,26 @@ class Bounds : public IRVisitor {
                                        !b_interval.min.type().is_uint() &&
                                        can_prove(b_interval.min < 0 &&
                                                  b_interval.min > -t.bits())) {
-                                interval.min = a_interval.min >> abs(b_interval.min);
+                                if (a_interval.min.type().is_uint() || can_prove(a_interval.min >= 0)) {
+                                    interval.min = a_interval.min >> abs(b_interval.min);
+                                } else if (can_prove(a_interval.min < 0)) {
```

Comment:
Seems weird to prove a_interval.min < 0. That's just that it *could* be negative. Did you mean a_interval.max?

Reply:
If the value *can* be negative, then the smallest possible value that *could* be produced by the shift is negative, but you can't produce a negative number from `a << b` if `a >= 0`, so it doesn't matter what `a`'s upper bound is

Reply:
You can think of it as splitting `a_interval` into two ranges, negative and non-negative. The non-negative interval (if it exists), cannot produce a negative result, no matter what the right-hand side is. Only the negative interval could produce a negative value, and in this block, that negative interval would be `[a_interval.min, 0)`, whose most negative value depends on whether or not you can increase the magnitude of any value on that interval or not.

Reply:
IMHO this explanation is worth capturing in a comment in the code, since it apparently wasn't obvious to two code reviewers

Reply:
Sounds good. This code has moved a bit from some of Andrew's other comments, but a comment on this has been added in 1a8930e